src/SettingsManager.js

```diff
@@ -0,0 +1,122 @@
+'use strict'
+
+let equal = require('deep-equal');
+let deepcopy = require('deepcopy');
+
+import { logger, configureLogger } from './logger';
+import cache from './cache';
+import Config from './Config';
+let authDataManager = require('./authDataManager');
+
+
+const settingsCollectionName = '_ServerSettings';
+
+// visible locked settings
+const lockedSettings = [
+  'applicationId',
+  'masterKey',
+  'serverURL',
+  'collectionPrefix',
+  'enableConfigChanges',
+];
+// hidden locked settings
+const hiddenSettings = [
+  'filesController',
+  'pushController',
+  'loggerController',
+  'hooksController',
+  'userController',
+  'authDataManager',
+  'liveQueryController',
+  'settingsInitialized'
+];
+
+// settings defined in server configuration
+let definedSettings = {};
+
+// callbacks for specific setting changes
+let authChange = settings => settings.authDataManager = authDataManager(settings.ouath, settings.enableAnonymousUsers);
+let onChange = {
+  logLevel: settings => configureLogger({ level: settings.logLevel }),
+  verbose: settings => {
+    settings.logLevel = 'silly';
+    configureLogger({ level: settings.logLevel });
+  },
+  oauth: authChange,
+  enableAnonymousUsers: authChange,
+  sessionLength: settings => settings.sessionLength = Number(settings.sessionLength),
+}
+
+export default function SettingsManager(appId) {
+  return {
+    pull: _ => {
+      return getSettingsCollection(appId)
+        .then(coll => coll.find({ applicationId: appId }, { limit: 1 }))
+        .then(results => {
+          let settings = results && results.length && results[0];
+          if (settings) delete settings._id;
+          logger.verbose('Pulled settings: ' + JSON.stringify(settings, null, 2));
+          return settings;
+        });
+    },
+
+    push: settings => {
+      settings = deepcopy(settings);
+      settings.applicationId = appId;
+      return getSettingsCollection(appId)
+        .then(coll => coll.upsertOne({applicationId: appId}, { $set: settings }))
+        .then(_ => {
+          logger.verbose('Pushed settings: ' + JSON.stringify(settings, null, 2));
+        });
+    },
+
+    updateCache: (updates = {}) => {
+      updates = Object.keys(updates)
+        .filter(update => !equal(updates[update], cache.apps.get(appId)[update]) && !isLocked(update) && !isDefined(update))
+        .reduce((filtered, update) => {
+          filtered[update] = updates[update];
+          return filtered;
+        }, {});
+
+      Object.keys(updates).forEach(setting => {
+        var config = cache.apps.get(appId);
+        logger.info(`Setting '${setting}' updated from '${config[setting]}' to '${updates[setting]}'`);
+        config[setting] = updates[setting];
+        if (onChange[setting]) onChange[setting](cache.apps.get(appId))
+      });
+      return updates;
+    },
+
+    getVisible: _ => {
+      let settings = cache.apps.get(appId);
+
+      let settingsString = JSON.stringify(settings, (k, v) => {
```

Comment:
I think that stringify + parse is a bit simpler because I need the deep replacer functionality to set values to null so that they appear in the GET response.  _.pick doesn't recurse deeply which causes some problems.  For example, customPages.invalidLink wouldn't show up in a GET if it was an undefined property.

I could implement a custom utility function for deep replacer functionality, but I don't think that it's worth it when getVisible is only called in the constructor and GET to settings router.